File: python_client/wisard_client.py
import asyncio
import gzip
import json
import time
from io import BytesIO

import httpx
import numpy as np


def load_mnist(prefix, folder):
    data_buffer = gzip.open(folder + prefix + '-images-idx3-ubyte.gz')
    data_buffer = data_buffer.read()
    int_type = np.dtype('int32').newbyteorder('>')
    metadata_bytes = 4 * int_type.itemsize

    data = np.frombuffer(data_buffer, dtype='ubyte')
    magic_bytes, n_images, width, height = np.frombuffer(
        data[:metadata_bytes].tobytes(), int_type)
    data = data[metadata_bytes:].astype(dtype='ubyte').reshape(
        [n_images, width * height])

    label_buffer = gzip.open(folder + prefix + '-labels-idx1-ubyte.gz')
    label_buffer = label_buffer.read()
    labels = np.frombuffer(label_buffer, dtype='ubyte')[2 * int_type.itemsize:]

    return data, labels


async def async_bytes(payload):
    yield payload


async def train_image(client, query, payload):
    await client.post("http://localhost:8080/train",
                      params=query,
                      data=async_bytes(payload))


async def classify_image(client, payload):
    resp = await client.post("http://localhost:8080/classify",
                             data=async_bytes(payload))
    r = resp.json()
    return r['label']


async def main():

    print("Pythonesque WiSARD - MNIST 🐍🐍🐍")
    async with httpx.AsyncClient() as client:
        # Get info
        r = await client.get("http://localhost:8080/info")
        wis_info = r.json()

        print("Number of hashtables: {}".format(wis_info['hashtables']))
        print("Address size: {}".format(wis_info['addresses']))
        print("Bleaching: {}".format(wis_info['bleach']))

    print("\n-----------------\nTraining\n-----------------")
    start_time = time.time()
    training_images, training_labels = load_mnist("train", "data/mnist/")
    print("Training data has {} images".format(len(training_labels)))
    print("Parsing the training dataset took: {:.0f} milliseconds".format(
        1000 * (time.time() - start_time)))

    start_time = time.time()
    async with httpx.AsyncClient() as client:
        tasks = []
        for (img, label) in zip(training_images, training_labels):
            payload = bytes(img)
            query = {'label': str(label)}
            tasks.append(
                asyncio.ensure_future(train_image(client, query, payload)))
        await asyncio.gather(*tasks)
    print("Training took: {:.0f} milliseconds".format(
        1000 * (time.time() - start_time)))

    print("-----------------\nTesting\n-----------------")
    start_time = time.time()
    test_images, test_labels = load_mnist("t10k", "data/mnist/")
    print("Testing data has {} images".format(len(test_labels)))
    print("Parsing the test dataset took: {:.0f} milliseconds".format(
        1000 * (time.time() - start_time)))

    hit = 0
    count = 0
    start_time = time.time()
    async with httpx.AsyncClient() as client:
        tasks = []
        for img in test_images:
            payload = bytes(img)
            tasks.append(asyncio.ensure_future(classify_image(client,
                                                              payload)))
        responses = await asyncio.gather(*tasks)
        for (response, label) in zip(responses, test_labels):
            if response == str(label):
                hit = hit + 1
            count = count + 1

    print("Testing took: {:.0f} milliseconds".format(
        1000 * (time.time() - start_time)))

    print("Accuracy: {:.4f}".format(hit / count))


asyncio.run(main())

File: src/wisard_client.rs
use dataloaders::mnist;
use http_types::Body;
use serde::{Deserialize, Serialize};
use std::fs::File;
use std::time::Instant;

#[async_std::main]
async fn main() -> surf::Result<()> {
    println!("-----------------\nTesting\n-----------------");
    let now = Instant::now();
    let label_data =
        &mnist::MnistData::new(&(File::open("data/mnist/t10k-labels-idx1-ubyte.gz"))?)?;
    let images_data =
        &mnist::MnistData::new(&(File::open("data/mnist/t10k-images-idx3-ubyte.gz"))?)?;
    let mut images: Vec<Vec<u8>> = Vec::new();
    let image_shape = (images_data.sizes[1] * images_data.sizes[2]) as usize;

    for i in 0..images_data.sizes[0] as usize {
        let start = i * image_shape;
        let image_data = images_data.data[start..start + image_shape].to_vec();
        images.push(image_data);
    }
    let classifications: Vec<u8> = label_data.data.clone();

    println!("Testing data has {} images", classifications.len());
    println!(
        "Parsing the test dataset took: {} milliseconds",
        now.elapsed().as_millis()
    );
    let mut hit: u64 = 0;
    let mut count: u64 = 0;

    let now = Instant::now();

    for (image, classification) in images.into_iter().zip(classifications.into_iter()) {
        // let tuple: (String, f64, f64);
        let ClassifyResponse { label } = surf::post("http://localhost:8080/classify")
            .body(Body::from_bytes(image))
            .recv_json()
            .await?;
        // tuple = wis.classify(image);
        if label == classification.to_string() {
            hit += 1;
        }
        count += 1;
    }
    println!("Testing took: {} milliseconds", now.elapsed().as_millis());

    println!("Accuracy: {}", hit as f64 / count as f64);

    Ok(())
}

#[derive(Debug, Deserialize, Serialize)]
struct ClassifyResponse {
    label: String,
}
